checker/src/org/checkerframework/checker/guieffect/GuiEffectVisitor.java

```diff
@@ -111,6 +119,71 @@ public boolean isValidUse(
         return ret;
     }
 
+    // For assignments and local variable initialization:
+    // Check for @UI annotations on the lhs, use them to infer @UI types on lambda expressions in the rhs
```

Comment:
Sounds good. I took a look into doing that today, particularly in conjunction with doing bottom-up instead of top-down inference. There are two issues I am having with doing that:

- This means instantiating the `@PolyUI` effect of every method invocation (inside a `@PolyUI` lambda) visited by `GuiEffectTreeAnnotator`, so that we know which effect to propagate up to the lambda expression. Ideally, I would like to move the code from lines 215-235 of `GuiEffectVisitor` into a new method for `GuiEffectTypeFactory` that we can call in both places. An obstacle for this is the call to `visitorState.getMethodReceiver()` in line 224, which provides info I don't know how to get inside the factory.
- I am actually not sure how to annotate the lambda expression in the tree. Is there a reasonable way to get the `AnnotatedTypeMirror` of the lambda given just the `LambdaExpressionTree` or is the idea to keep track of the `AnnotatedTypeMirror` passed to `visitLambdaExpression` on the way down and use that inside `visitMethodInvocation`?

Reply:
- Your first point sounds like a reasonable refactoring.  If there's no equivalent ```visitorState``` in the TreeAnnotator, though, that is an issue.  (For those not carefully perusing the code, @lazaroclapp is proposing to move the code that performs the instantiation of polymorphic effects based on receiver types to a helper method where it could be used by both checking and the TreeAnnotator.  The ```visitor.getMethodReceiver()``` call retrieves the ```this``` type for implicit self-dispatch.)  ```getMethodReceiver()``` just retrieves state tracked by the ```BaseTypeVisitor```, which originates on the method type of the current method (https://github.com/typetools/checker-framework/blob/a03f6baf82af518b46c18f7f3765688b5767b821/framework/src/org/checkerframework/common/basetype/BaseTypeVisitor.java#L455, see call to ```setMethodReceiver```).  The refactored method could accept the implicit receiver type as an argument, and call sites could pass the result of ```visitorState.getMethodReceiver()``` or a separate lookup on the type of the enclosing method as needed.
- Shouldn't the second parameter to an override of ```TreeAnnotator.visitLambdaExpression``` be the corresponding ```AnnotatedTypeMirror```?